prime-angular-frontend/src/app/modules/site-registration/pages/device-provider-page/device-provider-page.component.ts

```diff
@@ -0,0 +1,176 @@
+
+import { CardListItem } from '@shared/components/card-list/card-list.component';
+import { FormatDatePipe } from '@shared/pipes/format-date.pipe';
+
+import { AbstractCommunitySiteRegistrationPage } from '@registration/shared/classes/abstract-community-site-registration-page.class';
+import { IndividualDeviceProvider } from '@registration/shared/models/individual-device-provider.model';
+import { Site } from '@registration/shared/models/site.model';
+import { SiteFormStateService } from '@registration/shared/services/site-form-state.service';
+import { SiteService } from '@registration/shared/services/site.service';
+import { SiteRoutes } from '@registration/site-registration.routes';
+import { DeviceProviderPageFormState } from './device-provider-page-form-state.class';
+
+@Component({
+  selector: 'app-device-provider-page',
+  templateUrl: './device-provider-page.component.html',
+  styleUrls: ['./device-provider-page.component.scss'],
+  providers: [FormatDatePipe]
+})
+export class DeviceProviderPageComponent extends AbstractCommunitySiteRegistrationPage implements OnInit {
+  public formState: DeviceProviderPageFormState;
+  public form: FormGroup;
+
+  public currentIndex: number;
+
+  public title: string;
+  public isCompleted: boolean;
+  public isEditing: boolean;
+  public maxDateOfBirth: moment.Moment;
+  public busy: Subscription;
+
+  public contactCardListItems: CardListItem[];
+
+  public routeUtils: RouteUtils;
+
+  private site: Site;
+
+  constructor(
+    protected fb: FormBuilder,
+    protected dialog: MatDialog,
+    protected formUtilsService: FormUtilsService,
+    protected siteService: SiteService,
+    protected siteFormStateService: SiteFormStateService,
+    protected siteResource: SiteResource,
+    private formatDatePipe: FormatDatePipe,
+    protected route: ActivatedRoute,
+    router: Router
+  ) {
+    super(dialog, formUtilsService, siteService, siteFormStateService, siteResource);
+
+    this.title = route.snapshot.data.title;
+    this.routeUtils = new RouteUtils(route, router, SiteRoutes.SITES);
+
+    // Must be 18 years of age or older
+    this.maxDateOfBirth = moment().subtract(MINIMUM_AGE, 'years');
+    this.currentIndex = -1;
+  }
+
+  public onBack() {
+    const nextRoute = (this.isCompleted)
+      ? SiteRoutes.SITE_REVIEW
+      : SiteRoutes.HOURS_OPERATION;
+
+    this.routeUtils.routeRelativeTo(nextRoute);
+  }
+
+  public onCancel() {
+    this.isEditing = false;
+  }
+
+  public onAdd(): void {
+    this.currentIndex = -1;
+    this.form = this.formState.buildIndividualDeviceProvider();
+    this.isEditing = true;
+  }
+
+  public onEdit(index: number): void {
+    this.currentIndex = index;
+    this.form = this.formState.at(index);
+    this.isEditing = true;
+  }
+
+  public onContinue(): void {
+    if (this.formUtilsService.checkValidity(this.form)) {
+      if (this.currentIndex !== -1) {
+        this.formState.individualDeviceProviders.at(this.currentIndex).patchValue(this.form);
```

Comment:
Can access the indexed FormGroup using `this.formState.at(index).patchValue(...)` or after renaming `this.formState.individualDeviceProviderAt(index).patchValue(...)`